src/operations/search/search.js

```diff
@@ -207,108 +208,162 @@ module.exports.search = async (requestInfo, args, resourceName, collection_name)
-        let cursorQuery = collection.find(query, options).maxTimeMS(maxMongoTimeMS);
-
         /**
-         * @type {int | null}
+         * queries for logging
+         * @type {*|*[]}
          */
-        let cursorBatchSize = null;
-        // set batch size if specified
-        if (env.MONGO_BATCH_SIZE || args['_cursorBatchSize']) {
-            cursorBatchSize = args['_cursorBatchSize'] ? parseInt(args['_cursorBatchSize']) : parseInt(env.MONGO_BATCH_SIZE);
-            if (cursorBatchSize > 0) {
-                cursorQuery = cursorQuery.batchSize(cursorBatchSize);
-            }
-        }
+        let originalQuery = deepcopy(query);
         /**
-         * mongo db cursor
-         * @type {Promise<Cursor<unknown>> | *}
+         * options for logging
+         * @type {*|*[]}
          */
-        let cursor = await pRetry(
-            async () =>
-                await cursorQuery,
-            {
-                retries: 5,
-                onFailedAttempt: async error => {
-                    let msg = `Search ${resourceName}/${JSON.stringify(args)} Retry Number: ${error.attemptNumber}: ${error.message}`;
-                    logError(user, msg);
-                    await logMessageToSlack(msg);
-                }
-            }
-        );
+        let originalOptions = deepcopy(options);
 
-        // find columns being queried and match them to an index
         /**
-         * which index hint to use (if any)
-         * @type {string|null}
+         * whether to use the two-step optimization
+         * In the two-step optimization we request the ids first and then request the documents for those ids
+         *  This can be faster in large tables as both queries can then be satisfied by indexes
+         * @type {boolean}
          */
-        let indexHint = null;
-        if (isTrue(env.SET_INDEX_HINTS) || args['_setIndexHint']) {
-            indexHint = findIndexForFields(mongoCollectionName, Array.from(columns));
-            if (indexHint) {
-                cursor = cursor.hint(indexHint);
-                logDebug(user, `Using index hint ${indexHint} for columns [${Array.from(columns).join(',')}]`);
+        const useTwoStepSearchOptimization = !args['_elements'] && !args['id']
+            && (isTrue(env.USE_TWO_STEP_SEARCH_OPTIMIZATION) || args['_useTwoStepOptimization']);
+        if (useTwoStepSearchOptimization) {
+            // first get just the ids
+            const projection = {};
+            projection['_id'] = 0;
+            projection['id'] = 1;
+            options['projection'] = projection;
+            originalQuery = [query];
+            originalOptions = [options];
+            let idResults = await collection.find(query, options).maxTimeMS(maxMongoTimeMS).toArray();
+            if (idResults.length > 0) {
+                // now get the documents for those ids.  We can clear all the other query parameters
+                query = {'id': {$in: idResults.map(r => r.id)}};
```

Comment:
The 'query' variable is declared in line 54 and initialize in line 75, but does not get used until line 215.  Could we move the declaration/initialization closer to the actual usage?